homeassistant/components/esphome/entry_data.py

```diff
@@ -125,18 +133,65 @@ async def async_update_static_infos(
         async_dispatcher_send(hass, signal, infos)
 
     @callback
-    def async_update_state(self, hass: HomeAssistant, state: EntityState) -> None:
+    def register_state_type(
+        self, state_type: type[EntityState], component_key: str
+    ) -> None:
+        """Register a state type."""
+        self.entity_state_type_to_component_key[state_type] = component_key
+        self.entity_state_type_name_to_component_key[
+            state_type.__name__
+        ] = component_key
+
+    @callback
+    def component_key_for_state(self, state: EntityState) -> str:
+        """Return the component key for a state."""
+        state_class_name = state.__class__.__name__
+
+        # We try to avoid doing a linear search for the
+        # the component key by looking up the class name
+        # in the mapping. If it's not there, we fall back
+        # to a linear search for the component key by doing
+        # an instance search over entity_state_type_to_component_key.
+        if state_class_name in self.entity_state_type_name_to_component_key:
+            return self.entity_state_type_name_to_component_key[state_class_name]
+
+        for (
+            state_type,
+            component_key,
+        ) in self.entity_state_type_to_component_key.items():
+            if isinstance(state, state_type):
+                return component_key
```

Comment:
Nothing is subclassing the EntityStates currently but we ever wanted to subclass them in the future we need an instance check.

Reply:
Since this is not currently used, why not raise here and leave a comment that subclasses destroys the fast path.

Reply:
I took it out with the other suggested change below